Network logging fixes

jukkar opened the pull request.

jukkar committed:
net: Clarify logging in networking code

Remove network specific default and max log level setting
and start to use the zephyr logging values for those.

Remove LOG_MODULE_REGISTER() from net_core.h and place the
calls into .c files. This is done in order to avoid weird
compiler errors in some cases and to make the code look similar
as other subsystems.

Fixes #11343
Fixes #11659

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: Convert to use generic logging macros

Use generic logging macros LOG_*() instead of NET_*() as the
latter are mostly meant for internal networking stack use.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

nordic-krch commented on subsys/net/ip/connection.c:
```diff
@@ -8,9 +8,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_CONN_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_conn, NET_LOG_LEVEL);
```
what about using `CONFIG_NET_CONN_LOG_LEVEL` directly?

jukkar replied:
I did it like this here are in other places you commented because LOG_LEVEL was not "found", I needed to check the current log level later in the code and did not want to use the the config symbol directly. Is it on purpose that LOG_LEVEL is not created by LOG_MODULE_REGISTER()?

nordic-krch commented on subsys/net/ip/nbr.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_IPV6_NBR_CACHE_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_nbr, NET_LOG_LEVEL);
```
use directly `CONFIG_NET_IPV6_NBR_CACHE_LOG_LEVEL`

nordic-krch commented on subsys/net/ip/net_if.c:
```diff
@@ -4,9 +4,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_IF_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_if, NET_LOG_LEVEL);
```
use directly `CONFIG_NET_IF_LOG_LEVEL`

nordic-krch commented on subsys/net/ip/net_pkt.c:
```diff
@@ -10,17 +10,19 @@
-#if defined(CONFIG_NET_DEBUG_NET_PKT_ALL)
 #undef NET_LOG_LEVEL
+#if defined(CONFIG_NET_DEBUG_NET_PKT_ALL)
 #define NET_LOG_LEVEL 5
```
this looks strange as logger highest level is 4.

jukkar replied:
Yes, unfortunately the max level is 4 as in this module we could use 5 :)
The code just sets the level to 5 if extra debugging is enabled, that way we do not need to check the CONFIG_NET_DEBUG_NET_PKT_ALL option when needed.

nordic-krch commented on subsys/net/ip/net_stats.c:
```diff
@@ -4,14 +4,15 @@
-#define LOG_MODULE_NAME net_stats
-
 #if defined(CONFIG_NET_STATISTICS_PERIODIC_OUTPUT)
 #define NET_LOG_LEVEL LOG_LEVEL_INF
```
imo, this could also be resolved to log level otherwise it overwrites kconfig setting which may be confusing.

nordic-krch replied:
yes, i was actually questioning presence of `CONFIG_NET_STATISTICS_PERIODIC_OUTPUT`. If we have logging level setting that might be enough. I was considering situation when someone (in the future when different log levels are present in that file) sets logs to DBG enables periodic output and wonders why DBG output is not seen.

nordic-krch commented on subsys/net/ip/route.c:
```diff
@@ -9,9 +9,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_ROUTE_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_route, NET_LOG_LEVEL);
```
use `CONFIG_NET_ROUTE_LOG_LEVEL`

nordic-krch commented on subsys/net/ip/rpl.c:
```diff
@@ -38,9 +38,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_RPL_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_rpl, NET_LOG_LEVEL);
```
use `CONFIG_NET_RPL_LOG_LEVEL`

nordic-krch commented on subsys/net/ip/tcp.c:
```diff
@@ -13,9 +13,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_TCP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_tcp, NET_LOG_LEVEL);
```
use `CONFIG_NET_TCP_LOG_LEVEL`

nordic-krch commented on subsys/net/l2/bluetooth/bluetooth.c:
```diff
@@ -4,9 +4,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_BT_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_bt, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_BT_LOG_LEVEL`

nordic-krch commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -4,9 +4,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_ETHERNET_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_ethernet, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_ETHERNET_LOG_LEVEL`

nordic-krch commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -4,9 +4,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_GPTP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_DECLARE(net_gptp, NET_LOG_LEVEL);
```
use `CONFIG_NET_GPTP_LOG_LEVEL`

nordic-krch commented on subsys/net/lib/app/client.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_APP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_DECLARE(net_app, NET_LOG_LEVEL);
```
use `CONFIG_NET_APP_LOG_LEVEL`

nordic-krch commented on subsys/net/lib/app/net_app.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_APP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_app, NET_LOG_LEVEL);
```
use `CONFIG_NET_APP_LOG_LEVEL`

nordic-krch commented on subsys/net/lib/app/server.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_APP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_DECLARE(net_app, NET_LOG_LEVEL);
```
use `CONFIG_NET_APP_LOG_LEVEL`

nordic-krch commented on subsys/net/lib/http/http_server.c:
```diff
@@ -4,13 +4,11 @@
 #define NET_LOG_LEVEL CONFIG_HTTP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_DECLARE(net_http, NET_LOG_LEVEL);
```
use `CONFIG_HTTP_LOG_LEVEL`

nordic-krch commented on subsys/net/lib/mqtt/mqtt.c:
```diff
@@ -4,8 +4,8 @@
-#define LOG_MODULE_NAME net_mqtt
-#define NET_LOG_LEVEL LOG_LEVEL_ERR
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_mqtt, LOG_LEVEL_ERR);
```
there is `CONFIG_MQTT_LOG_LEVEL` available.

jukkar replied:
That symbol is only for new MQTT impl. This dir is for the legacy library which does not have log level defined.

nordic-krch commented on tests/net/checksum_offload/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_ETHERNET_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_ETHERNET_LOG_LEVEL`

nordic-krch commented on tests/net/iface/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_IF_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_IF_LOG_LEVEL`

nordic-krch commented on tests/net/promiscuous/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_IF_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_IF_LOG_LEVEL`

nordic-krch commented on tests/net/ptp/clock/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_ETHERNET_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_ETHERNET_LOG_LEVEL`

nordic-krch commented on tests/net/rpl/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_RPL_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_RPL_LOG_LEVEL`

nordic-krch commented on tests/net/traffic_class/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_TC_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_TC_LOG_LEVEL`

nordic-krch commented on tests/net/tx_timestamp/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_ETHERNET_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_ETHERNET_LOG_LEVEL`

nordic-krch commented on tests/net/websocket/src/main.c:
```diff
@@ -11,10 +11,12 @@
 
-#include <ztest.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_WEBSOCKET_LOG_LEVEL`

nordic-krch commented on tests/net/vlan/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_L2_ETHERNET_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_L2_ETHERNET_LOG_LEVEL`

nordic-krch commented on tests/net/udp/src/main.c:
```diff
@@ -6,9 +6,11 @@
 #define NET_LOG_LEVEL CONFIG_NET_UDP_LOG_LEVEL
 
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
```
use `CONFIG_NET_UDP_LOG_LEVEL`

nordic-krch reviewed (changes requested):
Great work. Making this PR must have been fascinating, as fascinating as reviewing it.

rlubos commented:
Great work indeed, I believe it is a step into the right direction.

One thing that remains and I am concerned about is that we still have hard dependency to the logger module in the networking headers. It's still not possible to simply include networking header - we need to create a logger instance first (see for example `http_get` sample which creates logger only to avoid compilation errors).

The reason for that is that we use logger macros within some of the networking headers (mostly through the `NET_ASSERT`, there's also a lone `NET_INFO` call in `net_analyze_stack` in net_core.h. The most obvious way to solve that would be to move implementations into the source files, but perhaps there is a better way to solve that? @nordic-krch Any ideas?

Or we can just accept it as it is and live with that if we agree it's not a big deal. I just wanted to raise the issue to make sure we make a conscious decission here.

pfalcon commented:
> see for example http_get sample which creates logger only to avoid compilation errors

Well, that's awful. samples/net/sockets/http_get is a POSIX sample, and it should move into the direction of not having `#ifndef __ZEPHYR__` at all.

So, that thing is another regression introduced by new logging, and should be resolved.

nordic-krch commented:
@rlubos regarding logging in headers. There are 2 options:
- move those functions to c file
- declare module in header file (see https://docs.zephyrproject.org/latest/subsystems/logging/logger.html#usage) like:

```
static inline void foo(void)
{
     LOG_MODULE_DECLARE(foo, CONFIG_FOO_LOG_LEVEL);

     LOG_INF("foo");
}
```

jukkar commented:
> * move those functions to c file

This might be the best option here.

> * declare module in header file (see https://docs.zephyrproject.org/latest/subsystems/logging/logger.html#usage) like:

In practice this is not really feasible, as the header cannot really know the used log level in .c file. Unless we hard code it to some specific value. I will investigate this more too.

jukkar commented:
@nordic-krch I tried to reach you at IRC, my question is that if I do `LOG_MODULE_REGISTER(foobar, MY_LOG_LEVEL);` will the macro set LOG_LEVEL to value of MY_LOG_LEVEL

nordic-krch commented:
@jukkar i don't start irc these days but you can reach me on slack.

Macro do not set LOG_LEVEL.

jukkar commented:
> Macro do not set LOG_LEVEL.

Could it set it if LOG_LEVEL is not set already? The reason being that then each .c file would then know the current log level set. Also it would avoid the messy NET_LOG_LEVEL settings you see in my commits.

jukkar committed:
net: Move net_analyze_stack() function to net_core.c

No need to this function be inlined so move it to .c file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Various updates according to comments.

nordic-krch commented on samples/net/rpl_border_router/src/rpl.c:
```diff
@@ -4,8 +4,10 @@
+#define LOG_LEVEL LOG_LEVEL_DBG
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(net_rpl_br_sample, LOG_LEVEL);
```
Ok.

I've been looking a bit into how to get rid of those `if (LEVEL>=DBG)` since ideally compiler should take care of that. An option would be to mark helper function (like `net_addr_ntop`) as pure (`__attribute__ ((pure))`. That indicates that function has no effects except the return value. When function is marked that way then compiler will cut it once it is noticed that return value is not used (because LOG_DBG is cut by compiler). However, to make this happen you need to use return value. Example from`rpl.c`:
```
char me_mem[NET_IPV6_ADDR_LEN];
char other_mem[NET_IPV6_ADDR_LEN];
char * me, other;

me = net_addr_ntop(AF_INET6, &dio->dag_id, other_mem, NET_IPV6_ADDR_LEN);
other = net_addr_ntop(AF_INET6, &rpl.dag_id, me_mem, NET_IPV6_ADDR_LEN);

LOG_DBG(..., me, other);
```

what's your opinion?

nordic-krch replied:
One note: if attribute is added to `net_addr_ntop` then we need to ensure that return value is always used across whole zephyr because compiler will cut function if return value is not used.

jukkar replied:
> what's your opinion?

Something like that could be done but not for this patch. Marking functions as pure would also need some extra review about how the values are used as you mention.

nordic-krch replied:
sure, i agree that it's not in the scope of that patch

rlubos commented:
And what about the `NET_ASSERT` usage in the headers? It's the final dependency to the logger.  Shall we move functions calling it as well? I've identified a few headers that use it:
* `net_context.h` 
* `trickle.h`
* `net_if.h` 
* `net_offload.h`

jukkar commented:
> And what about the `NET_ASSERT` usage in the headers? It's the final dependency to the logger. Shall we move functions calling it as well?

I will convert the NET_ASSERT to use the zephyr standard assert.

jukkar committed:
net: core: Change NET_ASSERT*() to use zephyr standard assert

Avoid dependency to logging subsystem in NET_ASSERT*(), so
use standard Zephyr asserts instead.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos reviewed (approved):
Thank you, with recent changes issues that concerned me seems to be resolved. We can now safely use networking headers without worrying about the logger anymore :)

jukkar committed:
tests: net: tcp: Fix interface initialization

The network interface mac address was not initialized properly
which caused assert in net_if.c:2881 to fail.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: context: Disable assertion for net_context tests

The unit test gives net_context_get() an invalid protocol family
value which triggers an assert in net_context.c:160
This assert fails the test when it should not as we properly
check the return value. Fix this by disabling asserts for this
test.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed assertion issues with networking unit tests.

jukkar commented:
Fixed merge conflict and converted new sample/net/syslog_net to use generic macros.

pfalcon reviewed (approved):
Wonder why this isn't merged yet, giving +1 to that end. (Otherwise, just relying on detailed discussion and reviews done by others.)
